### ISSUE TITLE
(times 0) semantics 

### ISSUE THREAD

**Issue by stelleg** [GaloisInc/ivory#28]
```n  `times` loop``` claims to run a computation n times, but in the case of `n = 0`, because it runs from `(n-1)` down to `0`, `n` will underflow. Here's an example:

```haskell
loop :: Def ('[Ix 1000] :-> Uint32)
loop  = proc "loop" $ \ n -> body $ do
  a <- local (ival 0)
  n `times` \ _ix -> do
    a' <- deref a
    store a (a' + 1)
  result <- deref a
  ret result

loop_zero :: Def ('[] :-> Uint32)
loop_zero = proc "loop_zero" $ body $ (ret . safeCast) =<< call loop 0
```

One would expect `loop_zero` to return `0`, but it in fact will return `1000`. 

Should be a simple zero check to fix. Am happy to implement if issue accepted.

**Comment by leepike** [comment]
Thanks. Yes, please do; I'll merge.

**Comment by stelleg** [comment]
OK, I have two solutions. One is a quick fix that does an explicit zero check for the `for` and `times` functions. This has the downside that it adds a conditional every time these functions are used. I'm also not sure if you allow zero length arrays, but if you do then it could be a problem there as well.

The other solution that I've implemented is a more thorough solution that adds parameters for whether or not to be inclusive with the end value to the LoopIncr constructors.  This change required a fair amount of modification, and broke a couple of tests that I haven't had a chance to look into further.

Let me know which you'd prefer.

**Comment by leepike** [comment]
Can you post the more thorough solution? It is likely close to what we've been considering--- @jameysharp discovered an issue a while ago we still haven't fixed in which (0 :: Ix 1) passed to `for` results in a bug. We might as well fix this once and for all. Don't worry about breaking old tests.

**Comment by stelleg** [comment]
OK, I've created a pull request. 

I believe the issue existed for `for` and `times`, and should be fixed for both.

**Comment by jameysharp** [comment]
In case anyone wonders, discussion of this issue continued over on pull request #30.

**Comment by elliottt** [comment]
OK, I've implemented a few changes as of 9d2d186242a4cc8dc57c19c9cc7cda044d0b8acc to deal with this:
* The `loop` function in `Ivory.Language.Loop` now takes `IxRep` arguments for the bounds of the loop instead of `Ix n`, allowing for the indexes to be computed before the call to `loop`
* `for` and `times` now use `loop` directly, instead of via `upTo` or `downTo`
* `for` and `times` compute the loop bounds at `IxRep` instead of `Ix n`, which avoids the underflow in the case that `n` is `0`

Additionally, I've made `loop` no longer exported from the `Ivory.Language.Loop` module, and similarly cleaned up the export list for `Ivory.Language.Ix`.

The downside to this fix is that if you do give an argument of type `Ix 0` to `times`, you'll likely introduce either garbage data, or divide-by-zero. The fix for this is to always use the `div-zero` pass when generating code, which should catch cases like that.
